File: 3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python
#
# [3] 无重复字符的最长子串
#

# @lc code=start
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        k, res, m = -1, 0, {}
        for i, c in enumerate(s):
            if c in m and m[c] > k:
                k = m[c]
            else:
                res = max(res, i - k)
            m[c] = i
        return res
# @lc code=end

File: 3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        std::map<char, int> m;
        int res = 0;
        int last_idx = -1;
        for (int i = 0; i < s.size(); ++i) {
            auto chr = s[i];
            if (m.find(chr) != m.end()) {
                last_idx = max(last_idx, m[chr]);
            }
            res = max(res, i - last_idx);
            m[chr] = i;
        }
        return res;
    }
};
// @lc code=end

